rfcs/0100-sign-commits.md

```diff
@@ -0,0 +1,380 @@
+to a new commit.
+
+However, even if you sign your commits, Nix currently
+provides no way of automatically verifying said
+commits for trust.
+
+[Guix](https://guix.gnu.org/), a competitor to Nix, already
+support verifying commits through their own system, described
+in the following links:
+- **https://guix.gnu.org/en/blog/2020/securing-updates/**
+- https://github.com/guix-mirror/guix/commits/master
+- https://guix.gnu.org/manual/en/html_node/Commit-Access.html
+- https://guix.gnu.org/manual/en/html_node/Channel-Authentication.html
+
+In addition, most conventional Linux distributions already sign their
+binary packages. By adding this functionality to Nix, end-users of Nixpkgs
+and other Nix repositories can be sure that they're getting the correct
+source code, **regardless** of the channel through they received
+it.
+
+This RFC was motivated by work during Summer of Nix, specifically,
+the Libre SoC maintainer requested that source integrity be
+maintained.
+
+# Detailed design
+[design]: #detailed-design
+
+## Verifying trust
+[verifying trust]: #verifying-trust
+
+### The algorithm
+
+The design is inspired by what is described here: https://guix.gnu.org/en/blog/2020/securing-updates/.
+
+We have two inputs to consider, an old trusted commit,
+and a new potentially untrusted commit.
+
+In the context of Git, a parent is a direct parent in the DAG of commits.
+
+The old commit is always trusted.
+
+If a commit has a trusted parent, then the commit is trusted if
+it doesn't have a trusted parent with `/.well-known/authorized-committers.nix`,
+
+If a commit has a trusted parent,
+and is signed by a key listed in each of the trusted
+parents' `/.well-known/authorized-committers.nix`,
+and the key hasn't expired or been revoked, regardless
+of when the commit has made, then the commit is trusted.
+
+The reason the timestamp of the commit doesn't matter,
+is that it's not secure and can be forged.
+
+The reason that a commit only needs one trusted parent,
+instead of all parents being trusted like Guix, is that
+this allows merging non-trusted commits, while the merge
+commit itself would be trusted.
+
+### The format of `/.well-known/authorized-committers.nix`
+
+An example: `/.well-known/authorized-committers.nix`:
+```nix
+{
+  committers = {
+    jacob = "XXXX XXXX XXXX XXXX XXXX  XXXX XXXX XXXX XXXX XXXX";
+    jonathan = "YYYY YYYY YYYY YYYY YYYY  YYYY YYYY YYYY YYYY YYYY";
+  };
+}
+```
+
+Specifically, it will contain a Nix a set with a single attribute `committers`, which contains
+another set where names are mapped to PGP fingerprints. This is a set
+instead of a list to make it easy to get the fingerprint for a specific
+person.
+The fingerprints listed should be registered on at least https://keys.openpgp.org/.
+
+The file must contain no thunks, and will be a restricted subset of Nix,
+like the top-level of `flake.nix`.
+
+### The Nix interface
+
+A new command, `nix verify-git` will be added:
```

Comment:
Does this mean that nix would depend on gpg? After writing https://github.com/Mic92/sops-nix it was not a pleasant experience and Debian also tries to migrate away from this software for good reason: https://wiki.debian.org/Teams/Apt/Spec/AptSign
It's UX is quite horrible leading to the creation of age: https://github.com/FiloSottile/age

Reply:
I assume we'd depend on Sequoia internally.

Reply:
Ok. My points has been already mentioned before.